Sources/FoundationEssentials/Calendar/Calendar_Recurrence.swift

```diff
@@ -0,0 +1,835 @@
+                    // Skip every few iterations when an interval has been given
+                    if (iterations - 1) % recurrence.interval != 0 {
+                        continue
+                    }
+                    // If a range has been specified, we should skip a few extra 
+                    // occurrences until we reach the start date
+                    if let rangeLowerBoundInterval {
+                        if nextDate < rangeLowerBoundInterval.start {
+                            continue
+                        }
+                    }
+                    return nextDate
+                }
+                return nil
+            }
+            
+            /// Compute `currentGroup` by advancing by a frequency interval from
+            /// the most recent group
+            mutating func nextGroup() {
+                guard let date = self.nextBaseRecurrenceDate() else {
+                    finished = true
+                    return
+                }
+                
+                var dates: [Date] = [date]
+                 
+                // First expand the set of dates, and then filter it
+                for action in [ComponentAction.expand, ComponentAction.limit] {
+                    if monthAction == action {
+                        recurrence._expandOrLimitMonths(dates: &dates, anchor: date, action: action)
+                    }
+                    if weekAction == action, action == .expand {
+                        recurrence._expandWeeks(dates: &dates, anchor: date) 
+                    }
+                    if dayOfYearAction == action {
+                        recurrence._expandOrLimitDaysOfTheYear(dates: &dates, anchor: date, action: action)
+                    }
+                    if dayOfMonthAction == action {
+                        recurrence._expandOrLimitDaysOfTheMonth(dates: &dates, anchor: date, action: action)
+                    }
+                    if weekdayAction == action {
+                        recurrence._expandOrLimitWeekdays(dates: &dates, anchor: date, action: action)
+                    }
+                    if hourAction == action {
+                        recurrence._expandOrLimitTimeComponent(.hour, dates: &dates, anchor: date, action: action)
+                    }
+                    if minuteAction == action {
+                        recurrence._expandOrLimitTimeComponent(.minute, dates: &dates, anchor: date, action: action)
+                    }
+                    if secondAction == action {
+                        recurrence._expandOrLimitTimeComponent(.second, dates: &dates, anchor: date, action: action)
+                    }
+                }
+                
+                dates = dates.filter { $0 >= self.start }
+                
+                if case let .afterDate(limit) = recurrence.end._guts {
+                    let hadDates = !dates.isEmpty
+                    dates = dates.filter { $0 <= limit }
```

Comment:
I'm not sure if this is actionable, but it seems like a reason to specify a limit is to avoid the cost of finding those dates in the first place - not just to filter them out at the end. Is it possible to do that?